sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/EliminateNullAwareAntiJoin.scala

```diff
@@ -20,22 +20,50 @@ package org.apache.spark.sql.execution.adaptive
 import org.apache.spark.sql.catalyst.planning.ExtractSingleColumnNullAwareAntiJoin
 import org.apache.spark.sql.catalyst.plans.logical.{LocalRelation, LogicalPlan}
 import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.exchange.ShuffleExchangeExec
 import org.apache.spark.sql.execution.joins.EmptyHashedRelationWithAllNullKeys
+import org.apache.spark.sql.execution.metric.SQLShuffleWriteMetricsReporter.SHUFFLE_RECORDS_WRITTEN
 
 /**
- * This optimization rule detects and convert a NAAJ to an Empty LocalRelation
- * when buildSide is EmptyHashedRelationWithAllNullKeys.
+ * This optimization rule try to Eliminate NAAJ by following patterns.
+ * 1. Convert a NAAJ to an Empty LocalRelation when buildSide is EmptyHashedRelationWithAllNullKeys.
+ * 2. Convert a NAAJ to left plan when buildSide is empty.
  */
 object EliminateNullAwareAntiJoin extends Rule[LogicalPlan] {
 
-  private def canEliminate(plan: LogicalPlan): Boolean = plan match {
+  private def canConvertToEmptyLocalRelation(plan: LogicalPlan): Boolean = plan match {
     case LogicalQueryStage(_, stage: BroadcastQueryStageExec) if stage.resultOption.get().isDefined
       && stage.broadcast.relationFuture.get().value == EmptyHashedRelationWithAllNullKeys => true
+    case LogicalQueryStage(_, stage: ShuffleQueryStageExec) if stage.resultOption.get().isDefined =>
+      // Equivalent to EmptyHashedRelationWithAllNullKeys
+      val nullPartitionKeyMetrics = stage.shuffle.asInstanceOf[ShuffleExchangeExec]
```

Comment:
How accurate are metrics in spark ? Can they be double or under counted ?

Reply:
I am not sure, this nullPartitionKey metrics is kind of loose than other metrics, when it greater than 0 we can consider it effective, but what about under counted, maybe I can offline talk with @cloud-fan for some advise.

Reply:
just +1 with @agrawaldevesh comment. I kind of feel aggregated `SQLMetrics` can be approximated (e.g. with failed/speculated tasks), and it seems to be quite dangerous here to depend on the exact value of `SQLMetrics` (e.g. >0 or == 0 below), which can have correctness concern.

Reply:
Yeah the concern is that we get data corruption when say spark tasks fail etc and the counters are invalid. I don't think Counters in Spark are meant to be so reliable. 

One approach we could use -- haven't thought too deeply about it -- is to run some Spark computation on the shuffle outputs. Its like Spark using itself (just like how Delta uses Spark for metadata computation). I think it should be doable given the machinery of Encoders etc but I haven't done it :-P